docs/proposals/component-registration.md

```diff
@@ -0,0 +1,406 @@
+
+## Terminology
+
+A Kubernetes Cluster **Component**...
+
+- Is considered a part of the cluster
+- Must register (or be registered) with Kubernetes (apiserver) to join the cluster
+- Must know about Kubernetes (speak the API)
+- Must be deployed/managed by the cluster admin/operator (rather than a user)
+- Should expose a method for remotely probing the liveness and readiness of each instance (e.g. tcp/http/https).
+- May be scheduled/deployed on the cluster (primary components vs addon components)
+- May be required for the cluster to function (required vs optional)
+- May be added/removed from a running cluster (mutable cluster component membership over time)
+- May require admin API access (privileged behavior)
+- May consist of one or more instances
+- May be deployed on different (network accessible) machines
+- May be upgraded over time
+
+The **External Components** are components that are deployed and lifecycle managed by some external system.
+
+The **Addon Components** are components that are deployed and lifecycle managed by the Kubernetes to which they provide functionality. These components must be optional in order for them to be deployable by Kubernetes (at least until an HA bootstrapping method is developed). Kubernetes conformance tests currently require some of these (e.g. kube-dns), but they are not required for orchestration and scheduling functionality.
+
+The **Primary Components** (currently apiserver, controller-manager, and scheduler) are components that are required for orchestration and scheduling functionality. These components are currently all External Components but may become Addon Components, once an HA bootstrapping method is developed.
+
+The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
+
+**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
+
+**Etcd** is a consistent key value storage cluster. It should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself with Kubernetes but must be configured. In general, the apiserver attempts to hide/abstract Etcd from the rest of the components with its API.
+
+**ComponentStatuses** is a top-level Kubernetes API resource that exposes a `/componentstatuses` endpoint that shows component and Etcd status. It's unquestionably useful and used, but questionably named and implemented. However, it must continue to function (until v2) for reverse compatibility. See [Problems](#problems) for more details. `kubectl get componentstatuses` can be used from the CLI as a wrapper for the API..
```

Comment:
Isn't ServiceSelector sufficient on its own to identify external and
internal services (as per the other discussions in this thread?
ExternalServiceSelector is underneath the service abstraction (by virtue of
the target selector on the service).

On Oct 9, 2015, at 8:37 PM, Daniel Smith <notifications@github.com> wrote:

In docs/proposals/component-registration.md
<https://github.com/kubernetes/kubernetes/pull/13216#discussion_r41690151>:

> +- May be deployed on different (network accessible) machines
> +- May be upgraded over time
> +
> +The **External Components** are components that are deployed and lifecycle managed by some external system.
> +
> +The **Addon Components** are components that are deployed and lifecycle managed by the Kubernetes to which they provide functionality. These components must be optional in order for them to be deployable by Kubernetes (at least until an HA bootstrapping method is developed). Kubernetes conformance tests currently require some of these (e.g. kube-dns), but they are not required for orchestration and scheduling functionality.
> +
> +The **Primary Components** (currently apiserver, controller-manager, and scheduler) are components that are required for orchestration and scheduling functionality. These components are currently all External Components but may become Addon Components, once an HA bootstrapping method is developed.
> +
> +The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
> +
> +**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
> +
> +**Etcd** is a consistent key value storage cluster. It should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself with Kubernetes but must be configured. In general, the apiserver attempts to hide/abstract Etcd from the rest of the components with its API.
> +
> +**ComponentStatuses** is a top-level Kubernetes API resource that exposes a `/componentstatuses` endpoint that shows component and Etcd status. It's unquestionably useful and used, but questionably named and implemented. However, it must continue to function (until v2) for reverse compatibility. See [Problems](#problems) for more details. `kubectl get componentstatuses` can be used from the CLI as a wrapper for the API..

Yeah, I'm against it even staying as a top level resource, I do *not* like
having special snowflakes in our API. As a strawman, what I'd like to see
is something like this:

type ComponentGroup struct {
  Spec ComponentGroupSpec
  Status ComponentGroupStatus
}
type ComponentGroupSpec struct {
  ServiceSelector map[string]string
  ExternalServiceSelector map[string]string
}
type ComponentGroupStatus struct {
  Statuses []ServiceStatus
}

These objects would be creatable and gettable just like anything else.
(Users can use this for healthchecking their applications, not just the
cluster!) They'd be located at something like
/apis/extensions/v1beta1/componentgroups. At cluster setup, one of these
objects would be made in the kube-system namespace for tracking the
cluster's own components.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13216/files#r41690151>.